File: 2025-04-02/Happy_Number.java
package amazon;

public class Happy_Number {
    public static int getNext(int n) {
        int sum = 0;

        while(n > 0) {
            int d = n%10;
            sum += (d*d);
            n /= 10;
        }

        return sum;
    }
    public static boolean isHappy(int n) {
        int slow = getNext(n);
        int fast = getNext(getNext(n));

        while(slow != fast) {
            if(fast == 1) return true;
            slow = getNext(slow);
            fast = getNext(getNext(fast));
        }

        return slow == 1;
    }
    public static void main(String[] args) {
        System.out.println(isHappy(19));
    }
}

File: 2026-03-12/Happy_Number.py
from collections import deque
from typing import List


class Solution:
    def isHappy(self, n: int) -> bool:
        def getSum(n):
            sum = 0
            while n:
                tmp = n % 10
                sum += tmp ** 2
                n //= 10

            return sum

        slow, fast = n, getSum(n)
        while slow != fast:
            slow = getSum(slow)
            fast = getSum(getSum(fast))

        return True if fast == 1 else False


sol = Solution()
print(sol.isHappy(3))